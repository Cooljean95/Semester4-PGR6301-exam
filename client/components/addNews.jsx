import React, { useEffect, useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import { fetchJSON } from "./fetchJSON";

export function AddNews({ user, createNews }) {
  const [category, setCategory] = useState("Weather");
  const [title, setTitle] = useState("");
  const [text, setText] = useState("");
  const [article, setArticle] = useState([]);
  const navigate = useNavigate();

  useEffect(async () => {
    const res = await fetchJSON("/api/news");
    setArticle(res);
    console.log("gettitle", res);
  }, []);

  async function handleSubmit(event) {
    event.preventDefault();
    let uniq = true;

    for (let i = 0; i < article.length; i++) {
      if (article[i].title === title) {
        uniq = false;
      }
    }

    if (uniq) {
      const res = await fetch("/api/news", {
        method: "POST",
        headers: {
          "content-type": "application/json",
        },
        body: JSON.stringify({
          category: category,
          title: title,
          text: text,
          author: user,
        }),
      });
      for (let i = 0; i < 3; i++) {
        const sw = new WebSocket(window.location.origin.replace(/^http/, "ws"));
        const newsArticle = {
          author: user,
          category: category,
          title: title,
          text: text,
        };
        sw.onopen = () => sw.send(JSON.stringify(newsArticle));

        if (res.ok) {
          navigate("/frontPage");
        }
      }
    } else {
      alert("title already exist");
    }
    createNews({ title });
  }

  return (
    <div>
      <h1>Add new movie</h1>
      <div>
        <Link to={"/frontPage"}>Frontpage</Link>
      </div>
      <form onSubmit={handleSubmit}>
        <label>
          Category:
          <select
            required={true}
            onChange={(event) => setCategory(event.target.value)}
          >
            <option>Weather</option>
            <option>Sport</option>
            <option>Food</option>
            <option>War</option>
            <option>Crimes</option>
          </select>
          <label>
            Title:
            <input
              required={true}
              value={title}
              onChange={(e) => setTitle(e.target.value)}
            />
          </label>
        </label>
        <label>
          Text:
          <textarea
            required={true}
            value={text}
            onChange={(e) => setText(e.target.value)}
          />
        </label>
        <button>Add movie</button>
      </form>
    </div>
  );
}
