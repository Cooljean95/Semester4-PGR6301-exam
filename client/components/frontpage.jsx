import React, { useEffect, useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import { fetchJSON } from "./fetchJSON";

function ArticleLog({ article }) {
  return (
    <div>
      <strong>{article.author}</strong>
      <h1>{article.title}</h1>
      <h3>{article.category}</h3>
      <p>{article.text}</p>
    </div>
  );
}

export function Frontpage({ user, getUser }) {
  const navigate = useNavigate();
  const [news, setNews] = useState({});
  const [name, setName] = useState("");
  const [loading, setLoading] = useState(false);
  const [newArticle, setNewArticle] = useState({});

  async function logOut() {
    const res = await fetch("/api/loginApi", {
      method: "delete",
    });
    if (res.ok) {
      navigate("/");
    }
  }

  useEffect(() => {
    const sw = new WebSocket(window.location.origin.replace(/^http/, "ws"));
    sw.onmessage = (event) => {
      const val = JSON.parse(event.data);
      setNewArticle(val);
    };
  }, [newArticle]);

  useEffect(async () => {
    setLoading(true);
    await google();
    await getTitle();
    setLoading(false);
  }, []);

  async function google() {
    try {
      const res = await fetchJSON("/api/loginApi");
      const user = res.name;
      console.log("res", res);
      setName(user);
    } catch (error) {
      console.log(error);
      setName("");
      setLoading(false);
    }
  }

  async function getTitle() {
    const res = await fetchJSON("/api/news");
    setNews(res);
    setLoading(false);
  }

  if (loading) {
    return <div>Loading wait..</div>;
  }

  if (name) {
    getUser(name);
  }

  if (!name) {
    if (user) {
      return (
        <div className={"frontPage3"}>
          <header>
            <div>
              <Link to={"/profile"}>Profile</Link>
            </div>
            <div>
              <Link to={"/news"}>News</Link>
            </div>
            <div>
              <button onClick={logOut}>Log out</button>
            </div>
            <div>
              <h3>welcome: {user}</h3>
            </div>
          </header>
          <main>
            <div className={"mainDiv"}>
              <h1>News articles</h1>
              {Object.keys(news).map((i) => (
                <div key={i}>
                  <p>{news[i].title}</p>
                </div>
              ))}
            </div>
            <div className={"mainDivTwo"}>
              <h1>Articles</h1>
              <ArticleLog article={newArticle} />
            </div>
          </main>
        </div>
      );
    } else {
      return (
        <div className={"frontPage2"}>
          <header>
            <div>
              <Link to={"/news"}>News</Link>
            </div>
            <div>
              <button onClick={logOut}>Log out</button>
            </div>
            <div>
              <h3>welcome: Guest</h3>
            </div>
          </header>
          <main>
            <div className={"mainDiv"}>
              <h1>News articles</h1>
              {Object.keys(news).map((i) => (
                <div key={i}>
                  <p>{news[i].title}</p>
                </div>
              ))}
            </div>
            <div className={"mainDivTwo"}>
              <h1>Articles</h1>
              <ArticleLog article={newArticle} />
            </div>
          </main>
        </div>
      );
    }
  }

  return (
    <div className={"frontPage"}>
      <header>
        <div>
          <Link to={"/profile"}>Profile</Link>
        </div>
        <div>
          <Link to={"/news"}>News</Link>
        </div>
        <div>
          <Link to={"/news/new"}>Add News</Link>
        </div>
        <div>
          <Link to={"/news/update"}>Update News</Link>
        </div>
        <div>
          <button onClick={logOut}>Log out</button>
        </div>
        <div>
          <h3>welcome: {name}</h3>
        </div>
      </header>
      <main>
        <div className={"mainDiv"}>
          <h1>News articles</h1>
          {Object.keys(news).map((i) => (
            <div key={i}>
              <p>{news[i].title}</p>
            </div>
          ))}
        </div>
        <div className={"mainDivTwo"}>
          <h1>Articles</h1>
          <ArticleLog article={newArticle} />
        </div>
      </main>
    </div>
  );
}
