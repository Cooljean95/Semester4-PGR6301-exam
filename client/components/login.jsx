import { Link, useNavigate } from "react-router-dom";
import React, { useState } from "react";

export function Login({ onLogin }) {
  const navigate = useNavigate();
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");

  function handleSubmit(event) {
    event.preventDefault();

    if (username && password) {
      onLogin(username, password);
      navigate("/frontPage");
    } else if (!username && !password) {
      setPassword(undefined);
      setUsername(undefined);
      onLogin(username, password);
      navigate("/frontPage");
    } else {
      alert("You need to input both username and password, or none");
    }
  }

  return (
    <>
      <h1>Log inn</h1>
      <div>
        <form onSubmit={handleSubmit}>
          <p>
            <label>
              Username:
              <input
                value={username}
                onChange={(e) => setUsername(e.target.value)}
              />
            </label>
          </p>
          <p>
            <label>
              Password:
              <input
                value={password}
                onChange={(e) => setPassword(e.target.value)}
              />
            </label>
          </p>
          <button>Log inn/Guest</button>
        </form>
      </div>
      <div>
        <ul>
          <li>
            <Link to={"/loginApi"}>Google login</Link>
          </li>
        </ul>
      </div>
    </>
  );
}
