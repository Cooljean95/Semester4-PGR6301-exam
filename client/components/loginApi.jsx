import React, { useContext, useEffect } from "react";
import { fetchJSON } from "./fetchJSON";
import { LoginContextGoogle } from "../index";

export function LoginApi() {
  const { discovery_endpoint, client_id, response_type, scope } =
    useContext(LoginContextGoogle);
  useEffect(async () => {
    const { authorization_endpoint } = await fetchJSON(discovery_endpoint);
    const parameters = {
      response_type,
      client_id,
      scope,
      redirect_uri: window.location.origin + "/login/callback",
    };

    window.location.href =
      authorization_endpoint + "?" + new URLSearchParams(parameters);
  }, []);

  return (
    <>
      <h1>Please wait...</h1>
    </>
  );
}
