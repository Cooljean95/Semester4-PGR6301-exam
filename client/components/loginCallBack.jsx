import React, { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";

export function LoginCallBack() {
  const [error, setError] = useState();
  const navigate = useNavigate();

  useEffect(async () => {
    const { access_token } = Object.fromEntries(
      new URLSearchParams(window.location.hash.substring(1))
    );

    if (!access_token) {
      setError("Missing access token");
      return;
    }

    const res = await fetch("/api/loginApi", {
      method: "POST",
      headers: {
        "content-type": "application/json",
      },
      body: JSON.stringify({ access_token }),
    });
    if (res.ok) {
      navigate("/frontPage");
    } else {
      setError(`Failed api login ${res.status}, ${res.statusText}`);
    }
  }, []);

  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div>{error}</div>
      </div>
    );
  }

  return (
    <>
      <h1>Please wait....</h1>
    </>
  );
}
