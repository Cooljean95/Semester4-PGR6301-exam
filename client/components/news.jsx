import React from "react";
import { useLoader } from "./useLoader";
import { Link } from "react-router-dom";

function Card({ news: { category, title, text, author } }) {
  return (
    <>
      <h1>{title}</h1>
      <h3>{category}</h3>
      <h4>{author}</h4>
      <div>{text}</div>
    </>
  );
}

export function News({ listNews }) {
  const { loading, error, data } = useLoader(listNews);

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  return (
    <div>
      <h1>News Articles</h1>
      <header>
        <ul>
          <li>
            <Link to={"/frontPage"}>Go to Front page</Link>
          </li>
        </ul>
      </header>

      {data.map((news) => (
        <Card key={news.title} news={news} />
      ))}
    </div>
  );
}
