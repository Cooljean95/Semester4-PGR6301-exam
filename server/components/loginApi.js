import { Router } from "express";
import fetch from "node-fetch";
import bodyParser from "body-parser";
import { fetchJSON } from "./fetchJSON.js";

export function LoginApi() {
  const router = new Router();
  router.use(bodyParser.json());

  router.get("/", async (req, res) => {
    const { access_token } = req.signedCookies;
    const { userinfo_endpoint } = await fetchJSON(
      "https://accounts.google.com/.well-known/openid-configuration"
    );

    const userInfo = await fetch(userinfo_endpoint, {
      headers: {
        Authorization: `Bearer ${access_token}`,
      },
    });
    if (userInfo.ok) {
      res.json(await userInfo.json());
    } else {
      console.log(`Failed to fetch ${userInfo.status}, ${userInfo.statusText}`);
      res.sendStatus(500);
    }
  });

  router.post("/", (req, res) => {
    const { access_token } = req.body;
    res.cookie("access_token", access_token, { signed: true });
    res.sendStatus(200);
  });

  router.delete("/", (req, res) => {
    res.clearCookie("access_token");
    res.sendStatus(200);
  });

  return router;
}
