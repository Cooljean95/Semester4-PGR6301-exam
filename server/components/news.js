import { Router } from "express";

export function NewsApi(mongoDatabase) {
  const router = new Router();

  router.get("/", async (req, res) => {
    const news = await mongoDatabase
      .collection("News")
      .find()
      .map(({ category, title, text, author }) => ({
        category,
        title,
        text,
        author,
      }))
      .limit(100)
      .toArray();
    res.json(news);
  });

  router.post("/", (req, res) => {
    const { category, title, text, author } = req.body;
    mongoDatabase.collection("News").insertOne({
      category,
      title,
      text,
      author,
    });
    res.sendStatus(200);
  });

  return router;
}
