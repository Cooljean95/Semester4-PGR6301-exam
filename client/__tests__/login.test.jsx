import React from "react";
import ReactDOM from "react-dom";
import { Simulate } from "react-dom/test-utils";
import { BrowserRouter } from "react-router-dom";
import { Login } from "../components/login";

describe("test for login", () => {
  it("shows input", () => {
    const element = document.createElement("div");
    ReactDOM.render(
      <BrowserRouter>
        <Login />
      </BrowserRouter>,
      element
    );
    expect(element.innerHTML).toMatchSnapshot();
    expect(
      Array.from(element.querySelectorAll("form label")).map((e) => e.innerHTML)
    ).toEqual(['Username:<input value="">', 'Password:<input value="">']);
  });

  //Change doesn't wanna work
  /*
  it("adds user", () => {
    const createUser = jest.fn();
    const user = "Cool";
    const password = "12345";
    const element = document.createElement("div");
    ReactDOM.render(
      <BrowserRouter>
        <Login onLogin={createUser} />
      </BrowserRouter>,
      element
    );
    Simulate.change(element.querySelector("form input:nth-of-type(1)"), {
      target: { value: user },
    });
    Simulate.change(element.querySelector("form input:nth-of-type(2)"), {
      target: { value: password },
    });
    Simulate.submit(element.querySelector("form"));
    expect(createUser).toBeCalledWith({
      user,
      password,
    });
    expect(element.innerHTML).toMatchSnapshot();
  });

   */
});
