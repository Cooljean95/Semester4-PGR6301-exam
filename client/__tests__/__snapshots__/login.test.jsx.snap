// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add User shows input 1`] = `"<h1>Log inn</h1><div><form><p><label>Username:<input value=\\"\\"></label></p><p><label>Password:<input value=\\"\\"></label></p><button>Log inn/Guest</button></form></div><div><ul><li><a href=\\"/loginApi\\">Google login</a></li></ul></div>"`;

exports[`add User shows options 1`] = `"<h1>Log inn</h1><div><form><p><label>Username:<input value=\\"\\"></label></p><p><label>Password:<input value=\\"\\"></label></p><button>Log inn/Guest</button></form></div><div><ul><li><a href=\\"/loginApi\\">Google login</a></li></ul></div>"`;

exports[`test for login shows input 1`] = `"<h1>Log inn</h1><div><form><p><label>Username:<input value=\\"\\"></label></p><p><label>Password:<input value=\\"\\"></label></p><button>Log inn/Guest</button></form></div><div><ul><li><a href=\\"/loginApi\\">Google login</a></li></ul></div>"`;
